eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/CombinedEventCriteria.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2010-2024. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore;
+
+import jakarta.annotation.Nonnull;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+
+/**
+ * Implementation of the {@link EventCriteria} combining two different {@code EventCriteria} instances into a single
+ * {@code EventCriteria}.
+ *
+ * @author Steven van Beelen
+ * @since 5.0.0
+ */
+final class CombinedEventCriteria implements EventCriteria {
+
+    private final Set<String> types;
```

Comment:
Aren't criteria conceptually a multimap of `Class` -> `Set<Index>`, where each entry defines the type of event an a set of indices that must all match a given type?

Reply:
Yea that's completely right. The `Set<String>` is essentially a mirror of the types collection in the protobuf API for DCB with Axon Server.
I kinda left it as we aren't using the `types` collection **at all** in this PR so far.

Do you think we should keep/remove/adjust it for the time being?